### PR TITLE
[OpaquePtr] Disable OpenCL target extension types for typed pointers.

### PR DESCRIPTION
**Pull request by jcranmer-intel**

### CODE REVIEW

**Review by Fznamznon — COMMENTED** [review]
Could you please give some motivation for disabling?

**Comment by jcranmer-intel** [comment]
> Could you please give some motivation for disabling?

Disabling target extension types for typed pointers minimizes the impact on code that's going through existing opencl drivers using typed pointers. It's mostly necessary for opaque pointer support, so I don't think it's entirely unreasonable. If it's an issue, I can instead back that part of the change out and only focus on fixing the lit tests so that they pass with the underlying upstream change to enable target extension types.

**Comment by Fznamznon** [comment]
> Disabling target extension types for typed pointers minimizes the impact on code that's going through existing opencl drivers using typed pointers. It's mostly necessary for opaque pointer support, so I don't think it's entirely unreasonable.

That sounds reasonable to me. Could you please then add a test checking there is no extension types for typed pointers?

File: clang/lib/CodeGen/TargetInfo.cpp [hunk 1]
```diff
@@ -11227,6 +11227,8 @@ static llvm::Type *getSPIRVImageType(llvm::LLVMContext &Ctx, StringRef BaseType,
 llvm::Type *CommonSPIRTargetCodeGenInfo::getOpenCLType(CodeGenModule &CGM,
                                                        const Type *Ty) const {
   llvm::LLVMContext &Ctx = CGM.getLLVMContext();
+  if (Ctx.supportsTypedPointers())
```
Comment (Fznamznon) [inline clang/lib/CodeGen/TargetInfo.cpp:11230]: Does it make sense to do that only for SYCL?

**Comment by jcranmer-intel** [comment]
> > Disabling target extension types for typed pointers minimizes the impact on code that's going through existing opencl drivers using typed pointers. It's mostly necessary for opaque pointer support, so I don't think it's entirely unreasonable.
> 
> That sounds reasonable to me. Could you please then add a test checking there is no extension types for typed pointers?

There are already a few tests that check compilation in nonopaque mode, e.g., CodeGenSYCL/no_opaque_fpga_pipes.cpp, no_opaque_image_accessor.cpp, and no_opaque_sampler.cpp

**Comment by jcranmer-intel** [comment]
Friendly review ping?

File: clang/lib/CodeGen/TargetInfo.cpp (hunk 1 above)
Comment (jcranmer-intel) [reply clang/lib/CodeGen/TargetInfo.cpp:11230]: Running a lot of the OpenCL tests, not doing this for OpenCL causes far, far more issues.

**Comment by jcranmer-intel** [comment]
Review ping?

File: clang/lib/CodeGen/TargetInfo.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/CodeGen/TargetInfo.cpp:11230]: Is that because community doesn't support typed pointers anymore?

Comment (jcranmer-intel) [reply clang/lib/CodeGen/TargetInfo.cpp:11230]: More that the OpenCL CPU backend doesn't really support opaque pointers yet, nor the target extension types.

**Review by Fznamznon — APPROVED** [review]

**Pull request merged**